compiler-hacking.md

```diff
@@ -0,0 +1,464 @@
+
+    IntUnaryOperator adder(int offset)
+    static int adder$lambda$0(int offset, int x)
+
+When adder creates a `IntUnaryOperator` for `adder$lambda$0`, it must determine
+which values from the enclosing context must be forced into the returned
+lambda. These values are called captures. In this case, `this` isn't captured,
+so the target method is static, but if it accesses fields, then `this` must be
+captured.
+
+The Shesmu compiler must capture things in the same way. This gets much more
+complicated since Java's captures are either explicit references or,
+implicitly, `this`. Shesmu needs to capture:
+
+- `this`
+- any variables/constants that the olive mentions
+- the current olive line and column numbers
+- the signatures
+- the _stream_ variable, in some contexts
+
+Most of this capturing is necessary because of `Define`. Take the following:
+
+     Input shesmu;
+
+     Define foo(string x)
+       Where type = x;
+
+What does the signature of `foo` look like in bytecode?
+
+Let's assume it's `void foo(String x)` and build outward from there.
+
+First, it must take modify a stream of `ShesmuIntrospectionValue`:
+
+    Stream<ShesmuIntrospectionValue> foo(
+      Stream<ShesmuIntrospectionValue> input,
+      String x)
+
+Second, when updating the olive data flow information, it needs to associate
+that data with the calling olive, so it needs to take the line and column:
+
+    Stream<ShesmuIntrospectionValue> foo(
+      Stream<ShesmuIntrospectionValue> input,
+      int line,
+      int column,
+      String x)
+
+It's possible to use signatures inside the `Define`, but the variables that can
+be included in the signature can come from both the caller and callee:
+
+    Input shesmu;
+
+     Define foo(string x)
+       Where type = x; # A signature used here should have both `type`
+                         and `changed` in it
+     Olive
+       foo("hi")
+       Where now > changed + 3hour
+       Run ...;
+
+Therefore, signatures must be passed:
+
+    Stream<ShesmuIntrospectionValue> foo(
+      Stream<ShesmuIntrospectionValue> input,
+      int line,
+      int column,
+      long signature_count,
+      long signable_count,
+      Set<String> signature_names,
+      Set<String> signable_names,
+      Function<ShesmuIntrospectionValue, String> sha1_signature,
+      Function<ShesmuIntrospectionValue, String> json_signature,
+      String x)
+
+Currently, all signatures are passed, but that could be improved.
+
+In the case of the `Where` clause, `Predicate<ShesmuIntrospectionValue>` needs
+to be generated. It will have a different signature:
+
+    boolean check(
+      Instant now,
+      ShesmuIntrospectionValue streamValue)
+
+By Java convention, when making a lambda, the arguments to the lambda itself
+come after the captures. In the case of Shesmu's use of stream, this always the
+_stream value_ (_i.e._, the current item in the stream).
+
+When access in the stream variable `changed`, it will generate code like:
```

Comment:
Was this supposed to have quotes around it, or is 'changed' a special term in Shesmu?

Reply:
It's not a special term, but it's something that is provided by `Input shesmu;`.

Reply:
Should this be 'when accessing the stream variable 'changed'', then?